govcd/vm_affinity_rule.go

```diff
@@ -0,0 +1,397 @@
+			href = link.HREF
+			break
+		}
+	}
+	if href == "" {
+		return Task{}, fmt.Errorf("no link with VM affinity rule found in VDC %s", vdc.Vdc.Name)
+	}
+
+	task, err := vdc.client.ExecuteTaskRequest(href, http.MethodPost,
+		"application/vnd.vmware.vcloud.vmaffinityrule+xml", "error instantiating a new VM affinity rule: %s", affinityRuleDef)
+	if err != nil {
+		// if we get any error, we repeat the validation
+		// with a strict check on VM existence.
+		_, validationErr := validateAffinityRule(vdc.client, affinityRuleDef, true)
+		if validationErr != nil {
+			// If we get any error from the validation now, it should be an invalid VM,
+			// so we combine the original error with the validation error
+			return Task{}, fmt.Errorf("%s - %s", err, validationErr)
+		}
+		// If the validation error is nil, we return just the original error
+		return Task{}, err
+	}
+	return task, err
+}
+
+func (vdc *Vdc) CreateVmAffinityRule(affinityRuleDef *types.VmAffinityRule) (*VmAffinityRule, error) {
+
+	task, err := vdc.CreateVmAffinityRuleAsync(affinityRuleDef)
+	if err != nil {
+		return nil, err
+	}
+	// The rule ID is the ID of the task owner
+	ruleId := task.Task.Owner.ID
+
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, err
+	}
+
+	// Retrieving the newly created rule using the ID from the task
+	vmAffinityRule, err := vdc.GetVmAffinityRuleById(ruleId)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving VmAffinityRule %s using ID %s: %s", affinityRuleDef.Name, ruleId, err)
+	}
+	return vmAffinityRule, nil
+}
+
+func (vmar *VmAffinityRule) Delete() error {
+
+	if vmar == nil || vmar.VmAffinityRule == nil {
+		return fmt.Errorf("nil VM Affinity Rule passed for deletion")
+	}
+
+	if vmar.VmAffinityRule.HREF == "" {
+		return fmt.Errorf("VM Affinity Rule passed for deletion has no HREF")
+	}
+
+	deleteHref := vmar.VmAffinityRule.HREF
+	if vmar.VmAffinityRule.Link != nil {
+		for _, link := range vmar.VmAffinityRule.Link {
+			if link.Rel == "remove" {
+				deleteHref = link.HREF
+			}
+		}
+	}
+
+	deleteTask, err := vmar.client.ExecuteTaskRequest(deleteHref, http.MethodDelete,
+		"", "error removing VM Affinity Rule : %s", nil)
+	if err != nil {
+		return err
+	}
+	return deleteTask.WaitTaskCompletion()
+}
+
+func (vmar *VmAffinityRule) Update() error {
+	var err error
+	var affinityRuleDef *types.VmAffinityRule
+
+	if vmar == nil || vmar.VmAffinityRule == nil {
+		return fmt.Errorf("nil VM Affinity Rule passed for update")
+	}
+	if vmar.VmAffinityRule.HREF == "" {
+		return fmt.Errorf("VM Affinity Rule passed for update has no HREF")
+	}
+
+	// We validate the input, without a strict check on the VMs
+	affinityRuleDef, err = validateAffinityRule(vmar.client, vmar.VmAffinityRule, false)
+	if err != nil {
+		return fmt.Errorf("[Update] %s", err)
+	}
+	vmar.VmAffinityRule = affinityRuleDef
+
+	updateRef := vmar.VmAffinityRule.HREF
+	if vmar.VmAffinityRule.Link != nil {
```

Comment:
maybe possible to extract function for getting the link by  Rel type? and use in all functions

Reply:
done